### PR TITLE
Make modules & runners python2/3 compliant 

### PR DESCRIPTION
**Pull request by jschmid1**
Trying to achieve python2 and python3 compliance in all our modules/runners (including the stack.py) by leveraging python's 2to3 tool (with human supervision) and custom changes with a bit of refactoring.

All major changes and their root-causes have been documented in #781 

Known bugs:

* the cli currently fails to parse the sls files.
* after a automated reboot minions betimes drop out early
* probably some code in rgw which I have missed due to insufficient testing hardware (multiple rgw)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -720,7 +721,7 @@ install: copy-files setup.py
 	$(PKG_INSTALL) python-setuptools python-click
 	python setup.py install --root=$(DESTDIR)/
 
-rpm: pyc tarball test
```
Comment (jschmid1) [inline Makefile]: Temp fix, remove that later

**Comment by jschmid1** [comment]
please run basic suite

**Comment by jschmid1** [comment]
please run basic suite

**Comment by jan--f** [comment]
The Makefile is broken. the install task needs new deps [here](https://github.com/SUSE/DeepSea/pull/904/files?diff=split#diff-b67911656ef5d18c4ae36cb6741b7965R740).

should look like this:
```
install: pyc copy-files
```

**Comment by jan--f** [comment]
Also the spec file pulls in a bunch of python2 dependencies.

**Comment by jschmid1** [comment]
> Also the spec file pulls in a bunch of python2 dependencies.

That probably depends on your repo setup.

**Comment by jschmid1** [comment]
https://github.com/SUSE/DeepSea/pull/904#issuecomment-367657293

done

**Comment by kshtsk** [comment]
please run basic suite

**Comment by smithfarm** [comment]
I'm very curious to look at the teuthology log. The only way I can imagine it passing is that it already uses the SLE15 image?

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
@smithfarm there is some issues with the jobs, it's reporting passed even if it is failed, looking into fixing this

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
jenkins run health-ok suite

**Comment by kshtsk** [comment]
please run health-ok suite

**Comment by kshtsk** [comment]
please run health-ok suite

**Comment by kshtsk** [comment]
The health-ok or any other build will fail, do to: 
http://storage-ci.suse.de:8080/job/deepsea-teuthology-rpm-builder/6/console
```
[deepsea-teuthology-rpm-builder] $ /bin/bash /tmp/jenkins7951304477723838303.sh
sed "s/DEVVERSION/"0.9.1+49.g710eb922"/" setup.py.in > setup.py
#make sure to create bytecode with the correct version
find srv/ -name '*.py' -exec python3 -m py_compile {} \;
find: ‘python3’: No such file or directory
find: ‘python3’: No such file or directory
```
I'm going to install python3 on jenkins master slave (it is leap42.2) and rerun the tests.

**Comment by kshtsk** [comment]
please run health-ok suite

**Comment by kshtsk** [comment]
I can see numerous errors like this:
```
chown salt:salt /var/lib/jenkins/rpmbuild/BUILDROOT/deepsea-0.9.1+49.g710eb922-0.x86_64/srv/salt/ceph/admin/cache || true
chown: changing ownership of '/var/lib/jenkins/rpmbuild/BUILDROOT/deepsea-0.9.1+49.g710eb922-0.x86_64/srv/salt/ceph/admin/cache': Operation not permitted
```
I wonder why rpm make file is trying to change ownership to salt:salt on build system, where such user is absent.

**Comment by kshtsk** [comment]
please run rgw check

**Comment by jschmid1** [comment]
health-ok fails with: 

```
2018-02-22T22:06:04.866 INFO:teuthology.orchestra.run.target158069083021.stdout:Problem: nothing provides python3-click needed by deepsea-0.9.1+49.g710eb922-0.noarch
2018-02-22T22:06:04.866 INFO:teuthology.orchestra.run.target158069083021.stdout: Solution 1: do not install deepsea-0.9.1+49.g710eb922-0.noarch
2018-02-22T22:06:04.867 INFO:teuthology.orchestra.run.target158069083021.stdout: Solution 2: break deepsea-0.9.1+49.g710eb922-0.noarch by ignoring some of its dependencies

```

**Comment by smithfarm** [comment]
The failure is because of this:

```
  os_type: sle
  os_version: 12.3
```

As of this PR, all PRs targeting master will need to be tested on SLE15. PRs targeting SES5 branch should continue to be tested on SLE-12-SP3.

**Comment by smithfarm** [comment]
The SLE15/SES6 teuthology setup is still a WIP - I'll try to wrap that up today.

**Comment by jan--f** [comment]
There is still an issue with the make file. The pyc target needs to install python3-setuptools.

**Comment by jschmid1** [comment]
> There is still an issue with the makefile. The pyc target needs to install python3-setuptools.

For `make rpm` all deps get installed prior to the execution of `pyc` via the specfile.

I wonder what has changed with this PR. Did it work for you before? 

That's https://github.com/SUSE/DeepSea/pull/904/files#diff-b67911656ef5d18c4ae36cb6741b7965R755 the only change I made which should work when switching from py2->3

**Comment by smithfarm** [comment]
I'm close to getting a health-ok run on this PR, in a openSUSE Leap 15.0 environment. I got as far as Stage 2 before tripping over https://github.com/SUSE/DeepSea/issues/979

**Comment by smithfarm** [comment]
Even after working around #979, I ran into #981. Still, this clearly makes DeepSea work better in the target (py3-only) environment, so I'll merge this if someone else doesn't get to it first.

**Comment by jan--f** [comment]
> For make rpm all deps get installed prior to the execution of pyc via the specfile.

This is for make install. If we'd have the branch in the suse repo I'd be happy to push a fix.

**Comment by jschmid1** [comment]
> This is for make install. If we'd have the branch in the suse repo I'd be happy to push a fix.

done

**Comment by kshtsk** [comment]
please run health-ok suite

**Comment by kshtsk** [comment]
please run health-ok suite

**Comment by kshtsk** [comment]
@jschmid1 @jan--f Makefile should not try install any dependencies by default. You have to manage them via spec file for rpm.

@jschmid1 This patch breaks repo building step on PR check. Keep in mind, there is leap 42.3 installed, and it will never run with root permissions on storage-ci.

**Pull request merged**